index.html

```diff
@@ -1,390 +1,1451 @@
+      This example uses a `TextEncoder` to encode a {{DOMString}} as the
+      necessary {{Uint8Array}} for transmission.
+
+      <pre class="js">
+        const encoder = new TextEncoder();
+        const writer = port.writable.getWriter();
+        await writer.write(encoder.encode("PING"));
+        writer.releaseLock();
+      </pre>
+
+      When writing larger chunks it can be important to allow the port to apply
+      back pressure so that the serial transmitter does not get too far behind
+      sending data generated by the application. The
+      {{WritableStreamDefaultWriter/write()}} method returns a {{Promise}} which
+      resolves when data has been written. While having some data available in
+      the transmit buffer is important to maintain good throughput awaiting this
+      {{Promise}} before generating too many chunks of data is a good practice
+      to avoid excessive buffering.
+    </aside>
+
+    The {{SerialPort/writable}} getter steps are:
+
+    <ol>
+      <li>
+        If [=this=].{{[[writable]]}} is not `null`, return
+        [=this=].{{[[writable]]}}.
+      <li>If [=this=].{{[[state]]}} is not `"opened"`, return `null`.
+      <li>If [=this=].{{[[writeFatal]]}} is `true`, return `null`.
+      <li>Let |stream:WritableStream| be a [=new=] {{WritableStream}}.
+      <li>Let |writeAlgorithm| be the following steps:
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          <li>
+            [=In parallel=], run the following steps:
+            <ol>
+              <li>Invoke the operating system to write |chunk| to the port.
+              <li>
+                If the chunk was successfully written [=queue a task=] to
+                [=resolve=] |promise| with `undefined` on the [=serial port task
+                source=].
+              <li>
+                If an operating system error was encountered, [=reject=]
+                |promise| with an {{UnknownError}}.
+              <li>
+                If the port was disconnected, [=queue a task=] to [=reject=]
+                |promise| with a {{NetworkError}} on the [=serial port task
+                source=], set [=this=].{{[[writeFatal]]}} to `true`, and invoke
+                the steps to [=handle closing the writable stream=].
+            </ol>
+          <li>Return |promise|.
+        </ol>
+
+        <div class="note">
+          [[STREAMS]] specifies that |writeAlgorithm| will only be invoked after
+          the {{Promise}} returned by a previous invocation of this algorithm
+          has resolved. This is not ideal as it prevents implementations from
+          coalescing multiple chunks waiting in the {{WritableStream}}'s
+          internal queue into a single request to the operating system.
+
+          <p>
+          An alternative specification of this algorithm would define a software
+          buffer in which bytes are explicitly queued for transmission so that
+          multiple requests can be coalesced.
```

Comment:
Interesting to see you running into this. We've been aware of this problem on the streams side for some time. We have a very-opaque old issue about it https://github.com/whatwg/streams/issues/27 and a more helpful [section of the requirements document](https://github.com/whatwg/streams/blob/master/Requirements.md#for-certain-sinks-you-must-provide-a-means-of-performing-batch-writes-without-introducing-lag-in-the-normal-case) which we never did get around to solving.

/cc @ricea

Reply:
I've been waffling on how normatively to specify this but this is roughly how the implementation in Chromium works. A shared memory buffer of `bufferSize` bytes is allocated to accept chunks passed to `write()`. If the chunk fits entirely then `write()` resolves immediately. Otherwise it waits until sufficient data has been written to the port to make room for the chunk. The browser process is pulling data from this buffer in parallel to pass it to the operating system. This effectively allows the implementation to "look ahead" to future writes but has the side effect that `write()` can resolve "early". I've considered respecifying these algorithms in terms of operations on this buffer.

Reply:
I think it's probably OK to leave it as-is, but maybe add a note or even normative sentence softening

> Invoke the operating system to write chunk to the port

to explain that this might actually just queue the chunk for later writing, depending on implementation constraints, or OS drivers, or whatever. And that this does mean write() might resolve even if the serial port never actually sees the data, which could get delivered (or failed to be delivered) at a future date.

The Streams spec notes (in its non-normative developer documentation)

> Note that what "success" means is up to the underlying sink; it might indicate simply that the chunk has been accepted, and not necessarily that it is safely saved to its ultimate destination. 

and we added that specifically because of concerns like this, so this all makes sense to me.